### PR TITLE
REF: Consistent optional dependency handling

### PR DESCRIPTION
**Pull request by TomAugspurger**
Standardizes how we do our optional dependency checking.

1. Adds a new `import_optional_dependency` helper that handles optional importing, raising with a nice message, and warning / raising if the version is not correct.
2. Changes inline `try / except ImportError` to use `import_optional_dependency`
3. Documents all(?) optional dependencies with their minimum version.
4. Documents that this is how devs should do optional dependencies. 

```pytb
>>> pd.read_gbq('')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/taugspurger/sandbox/pandas/pandas/io/gbq.py", line 140, in read_gbq
    pandas_gbq = _try_import()
  File "/Users/taugspurger/sandbox/pandas/pandas/io/gbq.py", line 14, in _try_import
    extra=msg,
  File "/Users/taugspurger/sandbox/pandas/pandas/compat/_optional.py", line 82, in import_optional_dependency
    raise ImportError(message.format(name=name, extra=extra)) from None
ImportError: Missing optional dependency 'pandas_gbq'. pandas-gbq is required to load data from Google BigQuery. See the docs: https://pandas-gbq.readthedocs.io. Use pip or conda to install pandas_gbq.
```

### CODE REVIEW

File: pandas/compat/_optional.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    """
+    try:
+        module = importlib.import_module(name)
+    except ImportError:
+        if raise_on_missing:
+            raise ImportError(message.format(name=name, extra=extra)) from None
+        else:
+            return
+
+    minimum_version = VERSIONS.get(name)
```
Comment (TomAugspurger) [inline pandas/compat/_optional.py]: We should consider setting a minimum version for every optional dep, and just making this a `__getitem__`

File: pandas/core/missing.py [hunk 2]
```diff
@@ -275,12 +272,7 @@ def _interpolate_scipy_wrapper(x, y, new_x, method, fill_value=None,
             raise ImportError("Your version of Scipy does not support "
                               "PCHIP interpolation.")
     elif method == 'akima':
-        try:
```
Comment (TomAugspurger) [inline pandas/core/missing.py]: Akima1DInterpolater was added in SciPy 0.14, and we now require 0.19, so this additional check isn't needed anymore.

File: doc/source/install.rst [hunk 3]
```diff
@@ -252,87 +252,69 @@ Recommended Dependencies
...
+Pandas has many optional dependencies that are only used for specific methods.
+For example, :func:`pandas.read_hdf` requires the ``pytables`` package. If the
+optional dependency is not installed, pandas will raise an ``ImportError`` when
+the method requiring that dependency is called.
+
+========================= ================== =============================================================
+Dependency                Minimum Version    Notes
+========================= ================== =============================================================
+matplotlib                2.2.2              Visualization
+xarray                    0.8.2              pandas-like API for N-dimensional data
```
Comment (jreback) [inline doc/source/install.rst]: are these meant to be in some ordering?

I would maybe put them in sections: io, numeric, other i think

File: pandas/core/arrays/sparse.py [hunk 4]
```diff
@@ -2205,10 +2206,8 @@ def to_coo(self):
         float32. By numpy.find_common_type convention, mixing int64 and
         and uint64 will result in a float64 dtype.
         """
-        try:
-            from scipy.sparse import coo_matrix
-        except ImportError:
-            raise ImportError('Scipy is not installed')
+        import_optional_dependency("scipy")
```
Comment (jreback) [inline pandas/core/arrays/sparse.py]: wouldn't you write this as

```
scipy = import_optional_dependency('scipy')
from ....
```

to make this explicit?

**Review by jreback — CHANGES_REQUESTED** [review]
generally lgtm.

**Comment by WillAyd** [comment]
This is a great idea! Just to be clear did you purposely only select a subset of optional dependencies to update? I opened #26782 which could either be included here or done as a follow up

File: pandas/compat/_optional.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    try:
+        version = module.__version__
+    except AttributeError:
+        # xlrd uses a capitalized attribute name
+        version = module.__VERSION__
+    return version
+
+
+def import_optional_dependency(
+    name, extra="", raise_on_missing=True, on_version="raise"
```
Comment (WillAyd) [inline pandas/compat/_optional.py]: Can you add annotations here? Should be straightforward I think

File: doc/source/install.rst (hunk 3 above)
Comment (TomAugspurger) [reply doc/source/install.rst]: I roughly grouped the I/O, HTML parsers, and just threw the others wherever.

With headers, it doesn't look as nice I think. What do you think?

Screen Shot 2019-06-12 at 8 23 47 AM

File: pandas/core/arrays/sparse.py (hunk 4 above)
Comment (TomAugspurger) [reply pandas/core/arrays/sparse.py]: If you're using something from the top-level of the package, then yes. But we want at `scipy.sparse` which isn't imported by default. And we don't using anything from `scipy`'s top-level, so we would need

```
scipy = import_optional_dependency('scipy')  # noqa
```

**Comment by TomAugspurger** [comment]
> did you purposely only select a subset of optional dependencies to update?

I grepped for `except ImportError`s and did the ones that didn't look too complex (I think `sql.py` was a bit strange).

I seem to have just missed the xlrd one. Do you want me to handle it here?

**Comment by TomAugspurger** [comment]
Looks like I missed bottleneck too, sorry.

File: doc/source/install.rst (hunk 3 above)
Comment (jreback) [reply doc/source/install.rst]: yeah ok maybe alphabetize then?

**Comment by WillAyd** [comment]
> I seem to have just missed the xlrd one. Do you want me to handle it here?

Up to you

File: pandas/compat/_optional.py (hunk 1 above)
Comment (jreback) [reply pandas/compat/_optional.py]: I think you *do* this now? so can you change to ``[]``?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
small comment, otherwise lgtm.

File: pandas/compat/_optional.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/compat/_optional.py]: There are ones, like Jinja2, where we don't specify a minimum supported version.

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Great change @TomAugspurger !

File: pandas/compat/_optional.py [hunk 6]
```diff
@@ -36,11 +36,15 @@
...
 
 def _get_version(module: types.ModuleType) -> str:
-    try:
-        version = module.__version__
-    except AttributeError:
+    version = getattr(module, '__version__', None)
+    if version is None:
         # xlrd uses a capitalized attribute name
-        version = module.__VERSION__
+        version = getattr(module, '__VERSION__', None)
```
Comment (jreback) [inline pandas/compat/_optional.py]: should we not encode the '__version__' (or '__VERSION__') with the list of deps directly? (or have a way to override); in fact we *already* do this in pandas/util/_print_versions. so actually I would update that to use this list of deps (and also to add the version accessor)

**Comment by jreback** [comment]
actually @TomAugspurger see my comments. This duplicates a ton of code in _print_versions. Can you (or others) do a followup (or create an issue) about de-duplicating.

**Comment by WillAyd** [comment]
see #26814

**Comment by TomAugspurger** [comment]
FYI @pandas-dev/pandas-core you may want to glance through http://pandas-docs.github.io/pandas-docs-travis/development/contributing.html#optional-dependencies, when we get PRs adding a new optional dependency.

**Comment by jorisvandenbossche** [comment]
Nice!

**Pull request merged**